examples/computer_vision/byol_pytorch/model_def.py

```diff
@@ -0,0 +1,423 @@
+            # - Final results are reported on the test / validation data (TEST).
+            # We combine these two datasets, and then use a custom reducer to calculate the final result.
+            cls_val_dataset = build_dataset(
+                self.data_config,
+                DatasetSplit.CLS_VALIDATION,
+            )
+            test_dataset = build_dataset(self.data_config, DatasetSplit.TEST)
+            dataset = JointDataset([cls_val_dataset, test_dataset], ["lr_val", "test"])
+        else:
+            # When only reporting self-supervised loss, we just use CLS_VALIDATION.
+            dataset = build_dataset(
+                self.data_config,
+                DatasetSplit.CLS_VALIDATION,
+            )
+        return DataLoader(
+            dataset,
+            batch_size=self.context.get_per_slot_batch_size(),
+            pin_memory=True,
+            num_workers=self.data_config.num_workers,
+        )
+
+    def train_batch(
+        self, batch: TorchData, epoch_idx: int, batch_idx: int
+    ) -> Dict[str, torch.Tensor]:
+        if self.training_mode == TrainingMode.SELF_SUPERVISED:
+            return self._train_self_supervised_batch(batch, epoch_idx, batch_idx)
+        elif self.training_mode == TrainingMode.CLASSIFIER_ONLY:
+            return self._train_classifier_batch(batch, epoch_idx, batch_idx)
+        else:
+            # Should be unreachable.
+            raise Exception(f"Unknown TrainingMode {self.training_mode}.")
+
+    def _train_self_supervised_batch(
+        self, batch: TorchData, epoch_idx: int, batch_idx: int
+    ) -> Dict[str, torch.Tensor]:
+        imgs1, imgs2, _ = batch
+        set_learning_rate_warmup_cosine_anneal(
+            self.hparams,
+            self.byol_opt,
+            self.context.get_global_batch_size(),
+            batch_idx,
+            cast(int, self.context._epoch_len),
+        )
+        # Dummy shape needed to bypass check inside byol_model.forward.
+        x = AttrDict({"first": imgs1, "second": imgs2, "shape": [len(imgs1)]})
+        loss = self.byol_model.forward(x)
+        self.context.backward(loss)
+        self.context.step_optimizer(self.byol_opt)
+        set_ema_beta_cosine_anneal(
+            self.hparams, self.byol_model, batch_idx, cast(int, self.context._epoch_len)
+        )
+        # Note: EMA requires an aggregation_frequency of 1 or online network may not be synced
+        # across processes at this point.
+        self.byol_model.update_moving_average()
+        return {"loss": loss}
+
+    def _train_classifier_batch(
+        self, batch: TorchData, epoch_idx: int, batch_idx: int
+    ) -> Dict[str, torch.Tensor]:
+        imgs, labels = batch
+        embeddings = self.byol_model.forward(
+            imgs, return_embedding=True, return_projection=False
+        ).detach()
+        # May be called from validation, so need to enable gradients.
+        losses = {}
+        with torch.enable_grad():
```

Comment:
i'm not sure if this would have the same behavior as setting module to train for something like dropout.